x-pack/plugins/infra/server/lib/alerting/log_threshold/log_threshold_executor.ts

```diff
@@ -66,28 +73,97 @@ export const createLogThresholdExecutor = (alertUUID: string, libs: InfraBackend
     }
   };
 
+interface ProcessingParams<Results> {
+  results: Results;
+  params: LogDocumentCountAlertParams;
+  alertInstanceFactory: AlertExecutorOptions['services']['alertInstanceFactory'];
+  alertId: string;
+}
+
+interface GroupBucket {
+  key: Record<string, string>;
+  doc_count: number;
+}
+
+interface ReducedGroupBucket {
+  name: string;
+  documentCount: number;
+}
+
+const processUngroupedResults = ({
+  results,
+  params,
+  alertInstanceFactory,
+  alertId,
+}: ProcessingParams<InfraDatabaseSearchResponse<{}, Aggregations>>) => {
+  const { count, criteria } = params;
+
+  const alertInstance = alertInstanceFactory(`${alertId}-${UNGROUPED_FACTORY_KEY}`);
+  const documentCount = results.hits.total.value;
+
+  if (checkValueAgainstComparatorMap[count.comparator](documentCount, count.value)) {
+    alertInstance.scheduleActions(FIRED_ACTIONS.id, {
+      matchingDocuments: documentCount,
+      conditions: createConditionsMessage(criteria),
+      group: UNGROUPED_NAME,
+    });
+
+    alertInstance.replaceState({
+      alertState: AlertStates.ALERT,
+    });
+  } else {
+    alertInstance.replaceState({
+      alertState: AlertStates.OK,
+    });
+  }
+};
+
+const processGroupByResults = ({
+  results,
+  params,
+  alertInstanceFactory,
+  alertId,
+}: ProcessingParams<GroupBucket[]>) => {
+  const { count, criteria } = params;
+
+  const groupResults: ReducedGroupBucket[] = results.reduce((acc, groupBucket) => {
+    const groupName = Object.values(groupBucket.key).join(', ');
+    const groupResult = { name: groupName, documentCount: groupBucket.doc_count };
+    return [...acc, groupResult];
+  }, [] as ReducedGroupBucket[]);
```

Comment:
Explicitly mentioning the accumulator type generic argument of `reduce` could be a nice way to avoid repetition and type casting:

```suggestion
  const groupResults = results.reduce<ReducedGroupBucket[]>((acc, groupBucket) => {
    const groupName = Object.values(groupBucket.key).join(', ');
    const groupResult = { name: groupName, documentCount: groupBucket.doc_count };
    return [...acc, groupResult];
  }, []);
```

What do you think?